### PR TITLE
Enter and Connect commands

### LINKED ISSUE #317: [8pt] `cartridge enter/connect` command
I sure that we need an alternative for `tarantoolctl connect ./tmp/run/myapp.router.control`.

I suggest adding `cartridge connect INSTANCE_NAME` command.

### PR DESCRIPTION
**Pull request by dokshina**
I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Closes #317

### CODE REVIEW

File: cli/connect/console.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+	default:
+		return nil, fmt.Errorf("Unknown protocol: %s", console.protocol)
+	}
+
+	return evalFunc(console, funcBody, args...)
+
+}
+
+func loadHistory(console *Console) {
+	var err error
```
Comment (mRrvz) [inline cli/connect/console.go:158]: This variable declaration is not necessary

File: cli/connect/console.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+		return
+	}
+
+	console.historyFilePath = filepath.Join(homeDir, HistoryFileName)
+
+	console.historyLines, err = common.GetLastNLines(console.historyFilePath, MaxHistoryLines)
+	if err != nil {
+		return
+	}
+}
```
Comment (mRrvz) [inline cli/connect/console.go:171]: Maybe this function should return error or log it?

File: cli/connect/console.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+		console.input = ""
+		console.livePrefixEnabled = false
+	}
+
+	return executor, nil
+}
+
+func inputIsCompleted(input string, luaState *lua.LState) bool {
+	var err error
```
Comment (mRrvz) [inline cli/connect/console.go:242]: This variable declaration is not necessary

File: cli/connect/console.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	if ConsoleOutputMode(inWords[2]) == ConsoleYAMLOutput {
+		console.outputMode = ConsoleYAMLOutput
+		return
+	}
+
+	if ConsoleOutputMode(inWords[2]) == ConsoleLuaOutput {
+		console.outputMode = ConsoleLuaOutput
+		return
```
Comment (mRrvz) [inline cli/connect/console.go:289]: This  return is not necessary

File: cli/connect/console.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+			prompt.ASCIICodeBind{ // move to one word left
+				ASCIICode: []byte{0x1b, 0x62},
+				Fn: func(buf *prompt.Buffer) {
+					d := buf.Document()
+					wordLen := len([]rune(d.GetWordBeforeCursorWithSpace()))
+					buf.CursorLeft(wordLen)
+				},
+			},
+			prompt.ASCIICodeBind{ // move to one word right
+				ASCIICode: []byte{0x1b, 0x66},
```
Comment (mRrvz) [inline cli/connect/console.go:316]: Im not sure, but its not clear what these bytes are. Maybe add some comments?

File: cli/connect/console.go (hunk 1 above)
Comment (dokshina) [reply cli/connect/console.go:158]: This variable is needed here

```go
// 1
console.historyLines, err = common.GetLastNLines(console.historyFilePath, MaxHistoryLines)
```

In fact, it's created anyway here

```go
// 2
homeDir, err := common.GetHomeDir()
```

but I don't like when some line (1) depend on such variable that is created by some other line (2).

File: cli/connect/console.go (hunk 2 above)
Comment (dokshina) [reply cli/connect/console.go:171]: The commands history is a very good feature, but it's not required to work with interactive console.
So, error shouldn't be returned.

But you are right, I will add debug log messages on console creation.

File: cli/connect/console.go (hunk 4 above)
Comment (dokshina) [reply cli/connect/console.go:289]: This return makes this `if` block the same as `if` block above. Moreover, it allows to add any code below without changing this block and prevents future mistakes.

File: cli/connect/console.go (hunk 5 above)
Comment (dokshina) [reply cli/connect/console.go:316]: Moved to variables.

File: test/conftest.py [hunk 6]
```diff
@@ -22,11 +23,9 @@
...
 from utils import DEFAULT_RUN_DIR
 
-
-FILES_DIR = 'test/files'
-INIT_NO_CARTRIDGE_FILEPATH = os.path.join(FILES_DIR, 'init_no_cartridge.lua')
-INIT_IGNORE_SIGTERM_FILEPATH = os.path.join(FILES_DIR, 'init_ignore_sigterm.lua')
-INIT_ADMIN_FUNCS_FILEPATH = os.path.join(FILES_DIR, 'init_admin_funcs.lua')
+from project import INIT_NO_CARTRIDGE_FILEPATH
+from project import INIT_IGNORE_SIGTERM_FILEPATH
+from project import INIT_ADMIN_FUNCS_FILEPATH
```
Comment (mRrvz) [inline test/conftest.py:28]: I'm not sure, but PEP8 doesn't welcome such imports. 

```python
from project import (INIT_NO_CARTRIDGE_FILEPATH, 
     INIT_IGNORE_SIGTERM_FILEPATH, ...)
```

Maybe something like this?

File: test/integration/connect/conftest.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import os
+
+from project import Project
+from project import patch_cartridge_proc_titile
+from project import remove_dependency
+from project import replace_project_file
+from project import INIT_NO_CARTRIDGE_FILEPATH
+
+from utils import DEFAULT_CFG, DEFAULT_RPL_CFG
+from utils import ProjectWithTopology, Instance
```
Comment (mRrvz) [inline test/integration/connect/conftest.py:12]: why imports from ``project`` looks different from ``utils``? import from utils seems to me (and pep8) better

File: test/integration/connect/conftest.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+
+@pytest.fixture(scope="function")
+def project_with_instances(built_project, start_stop_cli, request):
+    cli = start_stop_cli
+    project = built_project
+
+    router = Instance('router', 8081, 'localhost:3301')
+    s1_master = Instance('s1-master', 8082, 'localhost:3302')
+    s1_replica = Instance('s1-replica', 8083, 'localhost:3303')
```
Comment (mRrvz) [inline test/integration/connect/conftest.py:64]: Just an idea (not sure): forward this constants to something like ``@dataclass``

File: test/integration/connect/utils.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        stdout=subprocess.PIPE,
+        stdin=subprocess.PIPE,
+    )
+
+    user_input = '\n'.join([c.command for c in commands]) + '\n'
+
+    out, _ = process.communicate(user_input.encode('utf-8'))
+    out = out.decode('utf-8')
+
+    print(out)
```
Comment (mRrvz) [inline test/integration/connect/utils.py:63]: This print can be removed

File: cli/commands/connect.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+		},
+		Args: cobra.MaximumNArgs(1),
+	}
+
+	rootCmd.AddCommand(connectCmd)
+
+	// FLAGS
+	configureFlags(connectCmd)
+
+	// username flag
```
Comment (olegrok) [inline cli/commands/connect.go:48]: username/password is not a flag, it's an option. Flag usually is a boolean value

File: cli/connect/binary.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/c-bata/go-prompt"
+	"github.com/tarantool/go-tarantool"
+)
+
+func binaryConnect(console *Console) error {
+	var err error
+
+	console.binaryConn, err = tarantool.Connect(console.connOpts.Address, tarantool.Opts{
+		User: console.connOpts.Username,
+		Pass: console.connOpts.Password,
```
Comment (olegrok) [inline cli/connect/binary.go:17]: AFAIR you use timeouts for unix sockets. I think you should use it for tcp sockets as well

File: cli/connect/connect.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	"os"
+
+	"github.com/tarantool/cartridge-cli/cli/common"
+	"github.com/tarantool/cartridge-cli/cli/context"
+	"github.com/tarantool/cartridge-cli/cli/project"
+	"github.com/tarantool/cartridge-cli/cli/running"
+)
+
+const (
+	tarantoolWordSeparators = "\t\r\n !\"#$%&'()*+,-/;<=>?@[\\]^`{|}~"
```
Comment (olegrok) [inline cli/connect/connect.go:14]: Please add a comment where you got such set of symbols.

File: cli/connect/console.go [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+
+		if !inputIsCompleted(console.input, console.luaState) {
+			console.livePrefixEnabled = true
+			return
+		}
+
+		handleSetOutput(console, console.input)
+
+		if console.historyFilePath != "" {
+			common.AppendStringToFile(console.historyFilePath, console.input+"\n")
```
Comment (olegrok) [inline cli/connect/console.go:243]: Looks like you every time reopen the same file. I'm not sure but looks like possible place where your prompt will lag.

File: cli/connect/console.go [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+
+		console.input = ""
+		console.livePrefixEnabled = false
+	}
+
+	return executor, nil
+}
+
+func inputIsCompleted(input string, luaState *lua.LState) bool {
+	if _, err := luaState.LoadString(input); err == nil || !strings.Contains(err.Error(), "at EOF") {
```
Comment (olegrok) [inline cli/connect/console.go:258]: `!strings.Contains(err.Error(), "at EOF")`

Please, describe it in comment

File: test/files/init_no_cartridge.lua [hunk 15]
```diff
@@ -18,3 +20,9 @@ if tnt_major < 2 or (tnt_major == 2 and tnt_minor < 2) then
         sock:sendto('unix/', notify_socket, 'READY=1')
     end
 end
+
+local sock_path = os.getenv('TARANTOOL_CONSOLE_SOCK')
+if sock_path ~= nil then
+    local console = require('console')
+    assert(pcall(console.listen, sock_path))
+end
```
Comment (olegrok) [inline test/files/init_no_cartridge.lua:28]: `box.cfg{listen = ...}` should be tested as well.

File: test/integration/connect/conftest.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    p = ProjectWithTopology(
+        cli,
+        project,
+        instances_list=[router, s1_master, s1_replica],
+    )
+
+    request.addfinalizer(lambda: p.stop())
+
+    p.start()
```
Comment (olegrok) [inline test/integration/connect/conftest.py:95]: Seems you need to put "stop" after "start".

If start for some reasons fails, you call finalizer and double fail occurs - you tries to stop instance that didn't start.

File: test/integration/replicasets/utils.py [hunk 17]
```diff
@@ -1,65 +1,3 @@
...
-        check_instances_running(self.cli, self.project, [name for name in self.instances], daemonized=True)
-
-    def stop(self):
-        self.cli.stop(self.project, force=True)
-        check_instances_stopped(self.cli, self.project, [name for name in self.instances])
-        self.cli.clean(self.project)
-
-        os.remove(os.path.join(self.project.path, DEFAULT_CFG))
-
-
```
Comment (olegrok) [inline test/integration/replicasets/utils.py:62]: This changes should be moved into separate patch.

File: test/conftest.py (hunk 6 above)
Comment (dokshina) [reply test/conftest.py:28]: I like your approach, but I suggest you to prettify our tests in separated patch. Please, create an issue and describe these problems.

File: cli/commands/connect.go (hunk 10 above)
Comment (dokshina) [reply cli/commands/connect.go:48]: I agree, but cobra framework calls all options flags. And I call them flags to prevent confuses.

File: test/integration/connect/utils.py (hunk 9 above)
Comment (dokshina) [reply test/integration/connect/utils.py:63]: This print is needed for `pytest -s`, it makes output more verbose. I will add a comment.

File: cli/connect/console.go (hunk 13 above)
Comment (dokshina) [reply cli/connect/console.go:243]: Tarantool `console` uses this file too. What about a case when it is used at the same time?

File: cli/connect/console.go (hunk 14 above)
Comment (dokshina) [reply cli/connect/console.go:258]: There are two comments below. Added one more with a link to `console` code.

File: test/integration/replicasets/utils.py (hunk 17 above)
Comment (dokshina) [reply test/integration/replicasets/utils.py:62]: Yes, I agree. But it's not enough. I would like test helpers and fixtures to be re-organized, since we added a lot of commands and tests become more complicated.

Comment (dokshina) [reply test/integration/replicasets/utils.py:62]: See https://github.com/tarantool/cartridge-cli/pull/421

File: cli/connect/console.go (hunk 13 above)
Comment (olegrok) [reply cli/connect/console.go:239]: In should be clarified additionally.

E.g. https://unix.stackexchange.com/questions/346062/concurrent-writing-to-a-log-file-from-many-processes

File: doc/connect.rst [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+    cartridge enter INSTANCE_NAME [flags]
+
+Flags:
+
+* ``--name`` - application name
+* ``--run-dir`` - directory where PID and socket files are stored
+  (defaults to ./tmp/run or "run-dir" in .cartridge.yml)
+
+Connects to instance via it's1 console socket placed in ``run-dir``.
```
Comment (olegrok) [inline doc/connect.rst:21]: `1` is seems to be misprint

File: test/integration/connect/utils.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        cmd,
+        cwd=project.path,
+        stderr=subprocess.STDOUT,
+        stdout=subprocess.PIPE,
+        stdin=subprocess.PIPE,
+    )
+
+    output, _ = process.communicate()
+    output = output.decode('utf-8')
+    print(output)
```
Comment (olegrok) [inline test/integration/connect/utils.py:110]: Do you really need print here?

Comment (olegrok) [reply test/integration/connect/utils.py:110]: `assert process.returncode == 1, 'Got: ' + output`

File: cli/connect/binary.go (hunk 11 above)
Comment (dokshina) [reply cli/connect/binary.go:17]: Added `common.EvalTarantoolConnTimeout`.

File: test/integration/connect/conftest.py (hunk 16 above)
Comment (dokshina) [reply test/integration/connect/conftest.py:95]: No, `start` starts more that one instance, in case when some of them failed (`start` failed), we need to stop other started instances.

File: cli/connect/console.go (hunk 13 above)
Comment (dokshina) [reply cli/connect/console.go:239]: Fixed.

File: test/integration/connect/utils.py (hunk 19 above)
Comment (dokshina) [reply test/integration/connect/utils.py:110]: Yea, I would like this output to be printed anyway. I often use it for local tesing.

File: cli/common/socket.go [hunk 20]
```diff
@@ -75,6 +75,18 @@ func ReadFromConn(conn net.Conn, endOfOutput string) ([]byte, error) {
...
+func endOfOutputFound(data []byte, endOfOutput string) bool {
+	endOfOutputLen := len(endOfOutput)
+	dataLen := len(data)
+
+	if dataLen < endOfOutputLen {
+		return false
+	}
+
+	endOfData := data[dataLen-endOfOutputLen:]
+	return string(endOfData) == endOfOutput
```
Comment (olegrok) [inline cli/common/socket.go:87]: Maybe hasPostfix is better here?

Comment (olegrok) [reply cli/common/socket.go:87]: Of course we don't want to cast whole data to string (in case it leads to copy of whole data buffer). But this "endswith" implementation isn't clear enough IMO.

File: cli/common/socket.go [hunk 21]
```diff
@@ -50,10 +50,10 @@ func ConnectToTarantoolSocket(socketPath string) (net.Conn, error) {
 }
 
 func ReadFromConn(conn net.Conn, endOfOutput string) ([]byte, error) {
-	tmp := make([]byte, 1024)
+	tmp := make([]byte, 1)
```
Comment (olegrok) [inline cli/common/socket.go:53]: Looks a bit strange. Why do you change it?

File: cli/common/socket.go [hunk 22]
```diff
@@ -20,7 +20,7 @@ const (
 )
```
Comment (olegrok) [inline cli/common/socket.go:20]: Would be great to move this changes into separate commit. With good description and brilliant commit message.

File: cli/connect/common.go [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+	Username string
+	Password string
+}
+
+type GetRawSuggestionsFunc func(console *Console, lastWord string) interface{}
+
+func getConnOpts(connString string, ctx *context.Ctx) (*ConnOpts, error) {
+	connOpts := ConnOpts{
+		Username: ctx.Connect.Username,
+		Password: ctx.Connect.Password,
```
Comment (olegrok) [inline cli/connect/common.go:32]: + `timeout`?

Comment (olegrok) [inline cli/connect/common.go:32]: \+ timeout?

Comment (dokshina) [reply cli/connect/common.go:32]: Timeout is a constant that doesn't depend on conn string.

File: cli/connect/plain_text.go [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+	case console.outputMode == ConsoleLuaOutput:
+		return strings.HasPrefix(dataPortion, tagPushPrefixLua)
+	}
+
+	return false
+}
+
+func getReadFromConnSetMode(console *Console, newOutputMode ConsoleOutputMode) ReadFromConnFunc {
+	readFromConnFunc := func(conn net.Conn) ([]byte, error) {
+		setModeResBytes := make([]byte, 1024)
```
Comment (olegrok) [inline cli/connect/plain_text.go:112]: Here you don't read byte by byte as here https://github.com/tarantool/cartridge-cli/pull/425. Is it ok?

Comment (dokshina) [reply cli/connect/plain_text.go:112]: Yes, here I read a result of `\set output <mode>` command. If everything before was read correclty, it should be `true` value or error message.

File: cli/common/socket.go (hunk 22 above)
Comment (dokshina) [reply cli/common/socket.go:20]: See #425

File: cli/connect/plain_text.go (hunk 24 above)
Comment (dokshina) [reply cli/connect/plain_text.go:112]: I will add a comment here, just because your Birthday tomorrow. I sure that my comment is a great gift.

Comment (olegrok) [reply cli/connect/plain_text.go:112]: > I sure that my comment is a great gift

The best gift is a gift that is made by own hands.

File: cli/connect/connect.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	if ctx.Project.Name == "" {
+		if ctx.Project.Name, err = project.DetectName(ctx.Running.AppDir); err != nil {
+			return fmt.Errorf(
+				"Failed to detect application name: %s. Please pass it explicitly via --name",
+				err,
+			)
+		}
+	}
```
Comment (olegrok) [inline cli/connect/connect.go:96]: There are at least three places with such code fragment. Consider some refactoring of FillCtx. Seems you need some GenericFillCtx to perform such checks.

Maybe not now but in this case please file an issue.

**Review by olegrok — APPROVED** [review]

File: cli/connect/connect.go (hunk 25 above)
Comment (dokshina) [reply cli/connect/connect.go:96]: See #427
